tpl/template_funcs.go

```diff
@@ -1392,6 +1392,41 @@ func replace(a, b, c interface{}) (string, error) {
 	return strings.Replace(aStr, bStr, cStr, -1), nil
 }
 
+// partialCache represents a cache of partials protected by a mutex.
+type partialCache struct {
+	mu sync.RWMutex
+	p  map[string]template.HTML
+}
+
+// Get retrieves partial output from the cache based upon the partial name.
+// If the partial is not found in the cache, the partial is rendered and added
+// to the cache.
+func (c *partialCache) Get(name string, contextList ...interface{}) (p template.HTML) {
+	var ok bool
+
+	if p, ok = c.get(name); !ok {
+		p = partial(name, contextList...)
+		c.set(name, p)
```

Comment:
It may not have much practical difference in this case, but this lock construction with a `set` and `get` method makes it impossible to get the locking right -- and since people are going to copy this code into their university coding assignments, we might as well do it right.

* The creation of the partial isn't protected by a lock, so there will be multiple invocations of the same `partial`
* Also it would be good to double-check the map once you get the write lock; the partial may have been added in the open window between unlock and lock.